src/classes/RD2_ValidationService_TEST.cls

```diff
@@ -0,0 +1,198 @@
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        npe03__Recurring_Donation__c rd = new TEST_RecurringDonationBuilder()
+            .build();
+
+        try {
+            insert rd;
+        } catch (Exception e) {
+            errMessage = e.getMessage();
+        }
+        System.assert(errMessage.contains(System.Label.RD_DonorIsRequired));
+
+    }
+
+    /***
+    * @description Verifies if it displays error message when there is no contact for Household account
+    */
+    @isTest
+    private static void shouldFailWhenNoContactIsSpecifiedForHHAccount() {
+        String errMessage = '';
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        Account acc = getHHAccount();
+        insert acc;
+
+        npe03__Recurring_Donation__c rd = new TEST_RecurringDonationBuilder()
+            .withAccount(acc.Id)
+            .build();
+        try {
+            insert rd;
+        } catch (Exception e) {
+            errMessage = e.getMessage();
+        }
+
+        System.assert(errMessage.contains(System.Label.RD_ContactMustBelongToAccount));
+
+    }
+
+    /***
+    * @description Verifies if it displays error message when the contact do not belong to
+        the selected HouseHold account.
+    */
+    @isTest
+    private static void shouldFailWhenContactDoesNotBelongToHHAccount() {
+        String errMessage = '';
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        Account acc = getHHAccount();
+        insert acc;
+
+        Contact con = UTIL_UnitTestData_TEST.getContact();
+        insert con;
+
+        npe03__Recurring_Donation__c rd = new TEST_RecurringDonationBuilder()
+            .withAccount(acc.Id)
+            .withContact(con.Id)
+            .build();
+        try {
+            insert rd;
+        } catch (Exception e) {
+            errMessage = e.getMessage();
+        }
+        System.assert(errMessage.contains(System.Label.RD_ContactMustBelongToAccount));
+    }
+
+    /**
+    * @description Should fail when there is an invalid contact for the account in the
+        Recurring Donations list.
+    */
+    @isTest
+    private static void shouldFailOnlyInvalidRecurringDonations() {
+
+        String errMessage ='';
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        List<Contact> contacts = new List<Contact>{
+            UTIL_UnitTestData_TEST.getContact(),
+            UTIL_UnitTestData_TEST.getContact()
+        };
+        insert contacts;
+
+        contacts = [SELECT AccountId FROM Contact WHERE Id IN :contacts];
+
+        List<npe03__Recurring_Donation__c> rds = new TEST_RecurringDonationBuilder()
+        .withAccount(contacts[0].AccountId)//only RD for one Contact will be valid
+        .withAmount(100)
+        .build(contacts);
+
+        List<Database.SaveResult> results = Database.insert(rds, false);
+        ERR_Handler.Errors errors = ERR_Handler.getErrors(results, rds);
```

Comment:
Can you check if lines 135-140 can be replaced with `SaveResult` checks:
```
for (Integer i = 0; i < results.size(); i++) {
    Database.SaveResult result = results[i];
    if (result.isSuccess()) {
        System.assertEquals(contacts[0].Id, rd.npe03__Contact__c, 'Insert should succeed when Contact belongs to the Household.');
    } else {
         System.assertEquals(contacts[1].Id, rd.npe03__Contact__c, 'Insert should fail when Contact does not belong to the Household.');
    }
}
```

Reply:
> Can you check if lines 135-140 can be replaced with `SaveResult` checks:
> 
> ```
> for (Integer i = 0; i < results.size(); i++) {
>     Database.SaveResult result = results[i];
>     if (result.isSuccess()) {
>         System.assertEquals(contacts[0].Id, rd.npe03__Contact__c, 'Insert should succeed when Contact belongs to the Household.');
>     } else {
>          System.assertEquals(contacts[1].Id, rd.npe03__Contact__c, 'Insert should fail when Contact does not belong to the Household.');
>     }
> }
> ```

I am getting assertion error after updating as the above, checking on it why, here are the details:
System.AssertException: Assertion Failed: Insert should fail when Contact does not belong to the Household.: Expected: 0031D00000PB5zyQAD, Actual: 0031D00000PB5zxQAD

Reply:
@RadSFDeveloper , I got what is the problem here. When it fails, the RD might not have that contact, instead have to verify if the error exists or not. I will change the assert statement accordingly, for the else condition like this: 
else {
                System.assert(result.getErrors()[0].getMessage().contains(System.Label.RD_ContactMustBelongToAccount));
            }

Reply:
Sounds good, thanks @meena-sfdc !